### PR TITLE
Use refcounting for results

### PR DESCRIPTION
**Pull request by zenhack**
Stacked on top of #373.

---

There is no functional change here yet; this just factors the recounting
we were already doing into a package, and exposes the ability to
refcount to other parts of the code.

This is a stepping stone towards addressing https://github.com/capnproto/go-capnproto2/issues/349; the idea is that the
method receiver should get a reference to this too to keep it alive for
the right amount of time.

### CODE REVIEW

File: internal/rc/rc.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+		refcount: init,
+	}
+}
+
+func (rc *Releaser) Decr() {
+	newCount := atomic.AddInt32(&rc.refcount, -1)
+	if newCount == 0 {
+		rc.release()
+		rc.release = nil
+	}
```
Comment (lthibault) [inline internal/rc/rc.go:22]: Consider adding a test for a negative `newCount` and panicking if we over-release.

**Review by lthibault — CHANGES_REQUESTED** [review]
I think this is good, but I'd like to take one last look at it after #327 & #373 has been merged, and after this has been rebased.

One small suggestion inline.

File: internal/rc/rc.go (hunk 1 above)
Comment (zenhack) [reply internal/rc/rc.go:24]: Done.

**Comment by zenhack** [comment]
Merged #373 back into this.

**Review by lthibault — APPROVED** [review]
